### ISSUE TITLE
Add MP4Parser

### ISSUE THREAD

**Issue by tesshucom** [tesshucom/jpsonic#1368]
Currently, ``ffprobe`` is used for video analysis. This has the problem of slow speed and relatively high CPU load.
Virtually the standard format for videos is MP4. Create an MP4 Parser that targets MP4 and allows for faster analysis.

<details>
<summary>Rough mechanism</summary>

> Unlike music formats where tags are stored at the beginning of the file in most cases, videos that are supposed to be edited have a flexible tag storage position. Even with the same number of files, the analysis of the video will be much heavier than musics.

> The command processing time of ffprobe itself is almost instantaneous. However, ffprobe tends to have an extremely high load on the CPU because there are few IO restrictions. If we parallelize without thinking, it will overflow.

> While the actual ffprobe process is instantaneous, the overhead is huge in exec with Java, so it takes about the same amount of time for a video with the smallest file size or a large file. (1 command 2-3 seconds on a general PC)

> MP4 analysis with Java alone is possible by using Apache Tika without using ffprobe. By performing analysis while appropriately switching between this method and external commands, the total scanning time can be shortened without increasing the load on the CPU.
 - Tika's processing time depends on disk speed and file size, and its Correlation coefficient is almost 1. 
 - On the other hand, the processing time of ffprobe has nothing to do with disk speed and file size, and its correlation coefficient is almost 0. 
 - Simply put, Tika is faster for small files, such as Youtube's high-quality music videos. Beyond a certain file size, big videos like recorded TV shows are faster with ffprobe.

![image](https://user-images.githubusercontent.com/27724847/157058468-7a1ce190-9f66-453f-bb6c-00f0666eee12.png)

> From the execution lead time of Tika and ffprobe, it is possible to calculate "the branch point that Tika is faster than ffprobe" for each device. The logic of "use ffprobe for videos over ???Mb" can be achieved by incorporating statistical calculations. 

 - When measuring the execution time of ffprobe for files existing on the NAS, there is a tendency for the population to be two (because there are multiple bottlenecks on IO). Other than that, it has a relatively simple normal distribution.
 - Since the execution time of ffprobe does not have to be a strict standard.  To avoid heavy calculations, Jpsonic uses an average below the median to estimate the lead time for ffprobe.
 - In other words, instead of averaging all command execution time histories, a sample of the "relatively early completed process" population is used for estimation. The reason for filtering samples that took a relatively long time to run the ffprobe command is that it become noise when deciding whether to use ffprobe / Tika.
 - Analyzing large files is slower in Tika. It is an overprocess to process files in the range where it is better to use the ffprobe command with Tika. The total time will be longer than if you ran ffprobe alone. To avoid this and get good results, you should underestimate the estimated execution time of ffprobe.
 
> According to a simple test, it may be faster to have Tika and ffprobe in parallel, even with complex calculations. Especially fast if managing relatively small files on your local disk. When parsing a large file on NAS, ffprobe is adopted because it exceeds Tika's lead time in most cases. In other words, it's the same as before. 

However, ffprobe will be also parallelized in this version. It's either "fast" or "more faster" for the user. It can be announced to users that it is quite comfortable for small files such as Youtube's high quality music videos.

</details>
<details>
<summary>Supported fields</summary>

 - This is the specification of Tika. ``Parser for the MP4 media container format, as well as the older QuickTime format that MP4 is based on.`` Therefore, if it conforms to Tika's specifications, it will be a highly versatile interface and less likely to depend on a specific implementation.
 - The tag parser of ffprobe can get almost all items entered with ID3. Simply put, ** sort and music brainz are overkill as video tags. There is no problem omitting it.
 - Note the bitrate. If you consider using it with Tika, it will not be supported(There is no cross-platform extended parser). 
 - Allows users to choose between two methods, ffprobe and ffprobe&Tika. The advantage of ffprobe & Tika is that small files such as music videos can be analyzed faster. Also, even if ffprobe is not installed, Jpsonic alone can analyze files of a specific size or smaller.
 - Make sure there are no major issues without reading the bitrate. If there is no problem, there is no reason to force the use of ffprobe.
   - -> Confirmed. There are some implementations that require the existence of "duration". bitRate is not mandatory.

fields | name space | ffprobe | tika
-- | -- | -- | --
artist | dc:creator/xmpDM:artist | Y | Y
title | dc:title | Y | Y
height | tiff:ImageLength | Y | Y
width | tiff:ImageWidth | Y | Y
album | xmpDM:album | Y | Y
album artist | xmpDM:albumArtist | Y | Y
composer | xmpDM:composer | Y | Y
disc number | xmpDM:discNumber | Y | Y
duration | xmpDM:duration | Y | Y
genre | xmpDM:genre | Y | Y
release date | xmpDM:releaseDate | Y | Y
track number | xmpDM:trackNumber | Y | Y
album artist sort |   | Y | N
album sort |   | Y | N
artist sort |   | Y | N
*bitRate* | *ffprobe Only!*  | *Y* | *N*
composer sort |   | Y | N
music braing release id |   | Y | N
music braing track id |   | Y | N
title sort |   | Y | N

</details>
<details open>
<summary>Fixes</summary>

 - [x] Add statistical class for MP4 parse.
 - [x] Add a class that can parse MP4 while switching between ffprobe external command and Java internal analysis using statistics
   - [x] Externalize the processing that depends on ffprobe, in the newly created MPParser
   - [x] The externalized ffprobe class will be modified to be used in the previous video parser class as well.
       - The analysis field of the newly created parser is referred to the Quicktime namespace specification. For videos with a matching tag format, the value in the target field will be loaded. 
       - Unlike legacy implementations, there is automates testing on Win/Linux for all items.

A new class for MP4 is added (not yet used for processing).

___

#1376 will be prioritized for a while. Because there is doubt about the specifications of the embedded image of the video.

___

Modifications related to use with Tika have been postponed to v111.3.0.

#1445

</details>
